Improve UI routing

fbarl opened the pull request:
There is a lot to unpack in this PR, but the general goal was to move in the direction of React Router as much as possible (the remaining parts can be gradually completed as the rest of the app moves to React).

#### Major Changes

* Split `page` search param into separate routes (with some renamings) - resolves #2297
* Use absolute paths for all HTML / JS / CSS assets (https://github.com/tmrowco/electricitymap-contrib/pull/2317/commits/caa1d7cfe08f9648779bd21eed1f88d3f1e85dd2) - this transition was necessary for the routes other than `/` to work with React Router
* Moved completely to managing search params through the URL state, keeping their copy in Redux only for some observer logic in `main.js` that still needs to be migrated to React - this brings us much closer to addressing #2296
* One implication of the above is that the `solar` and `wind` search params _aren't stored and fetched from cookies anymore_ - if this is not desirable, there is a way we could reintroduce them, but I don't see a big value in it as the browser URL history would normally contain the most used config anyway

#### Minor Changes

* Use a `history` instance (passed to the React Router wrapper) for managing state history in favor of  `window.history` as a more idiomatic React Router solution before the whole UI is moved to React
* Move between pages with `<Link>` and `<Redirect>` (which is more idiomatic for React Router) instead of pushing to browser history everywhere except `main.js`
* Added route redirects when initially landing on a page not appropriate for the current view (mobile vs. desktop)
* Introduced `usePageViewsTracker` hook for GA tracking config updates
* Removed `pageToGoBackTo` from the Redux state in favor of explicit URL navigation
* Removed `useRemoteEndpoint` from the Redux state in favor of the URL search param state https://github.com/tmrowco/electricitymap-contrib/pull/2317/commits/ae56a1d7a11ae9590f7d5c06e580f80f5313337e
* Renamed `showPageState` -> `currentPage`, `customDate` -> `customDatetime` in Redux

#### Tested

* The assets (images) are still being loaded correctly on:
  * [x] `docker-compose build && docker-compose up` (static build)
  * [x] `docker-compose watch-en && yarn start` (webpack server)
  * [ ] Cordova app(s) - not sure how to test before merging :thinking: 
* Old routes redirect to new ones:
  * [x] `/` -> `/map`
  * [x] `/?page=highscore&remote=true` -> `/ranking?remote=true`
  * [x] `/?page=country&solar=true&countryCode=DE` -> `/zone/DE?solar=true`
* Redirects between mobile and desktop screens
  * [x] `/info?solar=true` -> `/map?solar=true` (on desktop screen)
  * [x] `/faq?wind=true` -> `/info?wind=true` (on mobile screen)
* Moving around the app in both mobile and desktop view using:
  * [x] Clicks
  * [x] Keyboard
  * [x] Browser navigation

#### Out of Scope

* Pull router `<Switch>` to the top of the React tree and use the `Main` layout as a page wrapper for the left panel layouts - this looks straightforward from here but there are some intricacies involed in making sure that the rendering performance is not compromised and that the errors are still locally scoped (so they don't break the rendering of the whole app, see [error boundaries](https://reactjs.org/docs/error-boundaries.html))
* Have a separate sets of routes for mobile and desktop modes - this would be easy once the above is done and it would improve layouts readability which currently intermix the two modes and most of the time only partly render their content
* Have all the page view tracking hooks be called from the newly introduced `usePageViewsTracker` - this could make for a nice small follow-up PR

corradio commented on web/src/main.js:
```diff
@@ -253,7 +244,6 @@ function updateCo2Scale() {
   if (typeof zoneMap !== 'undefined') zoneMap.setCo2color(co2ColorScale, theme);
 }
 
-d3.select('#checkbox-colorblind').node().checked = getState().application.colorBlindModeEnabled;
```
are you sure this can be removed?

fbarl replied:
Ah yes, I must have removed it by mistake :man_facepalming: 

Reverted in https://github.com/tmrowco/electricitymap-contrib/pull/2317/commits/e5d0ac88d8b48e76817b96219251624b2cca2497, good catch! :+1:

fbarl replied:
Actually it seems that there was a reason why I originally removed that line - other changes introduced in this PR make some parts of the DOM render conditionally so when initially landing on a page that doesn't contain this checkbox, we'd hit an error trying to set its state from here.

I just made a fix in https://github.com/tmrowco/electricitymap-contrib/pull/2317/commits/1ac5fa36585255dc7b31c405647ef441090990a3 by adding `checked` and `onChange` triggers directly to the checkbox element in React.

corradio replied:
much better!

corradio commented on web/src/reducers/index.js:
```diff
@@ -26,7 +26,7 @@ const initialApplicationState = {
   co2ColorbarValue: null,
   colorBlindModeEnabled: cookieGetBool('colorBlindModeEnabled', false),
```
is it only wind/solar where the cookie is not set/restored?

fbarl replied:
Yes, the reason being that `solar` and `wind` are search params reflected in the URL state while the other 3 bools for which we still use cookies (`colorBlindModeEnabled`, `brightModeEnabled`, `onboardingSeen`) currently exist _only_ in Redux. Two points though:

* If we really want to keep the cookies for `solar` and `wind`, I could re-add them - the mechanism would just have to be a bit different as we'd have to redirect the URL to reflect the cookie values and update the cookies on URL change
* IMO we should also move some of the 3 aformentioned bools from Redux to the URL state -as I suggested in #2314

corradio commented on web/src/selectors/index.js:
```diff
@@ -5,9 +5,10 @@ import {
 
 export function getSelectedZoneHistory(state) {
-  return state.data.histories[state.application.selectedZoneName] || [];
+  return state.data.histories[getZoneId()] || [];
```
wouldn't it make sense to have a `getZoneId(state)` selector inside this file, instead of using router.js which imports store.js and uses getState()?
I think it's more logic to keep all the redux state selectors in this current file, but I might be wrong.

fbarl replied:
You're right, there was something deeply wrong (and potentially buggy) with Redux selectors depending on the URL state helpers.

The reason I didn't want to support `getZoneId(state)` selector is because I wanted to get as close as possible to #2296 in this PR and that in particular means getting `zoneId` from the URL and not refer to its Redux state copy whenever possible.

I went to make all Redux selectors depend only on the Redux state in https://github.com/tmrowco/electricitymap-contrib/pull/2317/commits/69b73dd5e541c94e62d1dc91133afda844ef3e65, i.e I made them pure. I did that by wrapping them in _builder_ functions which accept all the dependencies passed to them straight from React components.

Please take a look and see what you think!

fbarl commented:
@corradio Thanks for the review, I think I addressed all your concerns, PTAL!

Also, I don't know if there's a nice way to test that the Cordova app still works with the new routes and absolute asset paths - would be nice to give it a quick smoke test before we merge (or at least cut a new release).

corradio commented on web/src/selectors/index.js:
```diff
@@ -6,44 +6,45 @@ import {
 
-export function getSelectedZoneHistory(state) {
-  return state.data.histories[state.application.selectedZoneName] || [];
+export function getZoneHistory(zoneId) {
```
If all selectors are "accessor" I would add a suffix to the function name to make it clear: `getZoneHistoryAccessor` or `getZoneHistorySelector`. That would make it clear that we're dealing with a higher-order function.
Also it's maybe worth renaming that file to redux.js to make it clear that this file is about redux selectors?

fbarl replied:
Yes, I agree with both of your points - I addressed them in https://github.com/tmrowco/electricitymap-contrib/pull/2317/commits/256eead0c5a372e88d0e83e747caeba24ed929e9 and https://github.com/tmrowco/electricitymap-contrib/pull/2317/commits/00d5965438a74008cc8b046ca00d94159362d8c9.

However, I noticed today that the performance in rendering graphs got much worse after I started using `useSelector` hook and after some research I figured the reason had to do with [shallow compare being replaced by reference comparison](https://react-redux.js.org/api/hooks#equality-comparisons-and-updates), which effectively broke exchange keys calculation memoization and consequently caused `AreaGraph`s to be rerendered continuously when hovering over them.

That and the increased layers of abstraction when using the selectors nudged me to simplify the whole thing further so I actually replaced all those selectors with hooks in https://github.com/tmrowco/electricitymap-contrib/pull/2317/commits/3ff75dbda908ac7d7d9d2e5a3518f75b7a835a05, that do a bit more than just getting the data from Redux - pulling the URL params and doing the memoization is now done from the inside which means the new hooks are much simpler to use in the components, e.g.:

```
const historyData = useCurrentZoneHistory();
```

WDYT?

corradio replied:
this makes a lot of sense. LGTM

corradio commented:
Have you tried building the cordova app by following the info in the README at https://github.com/tmrowco/electricitymap-contrib/tree/master/mobileapp ?

fbarl commented:
> Have you tried building the cordova app by following the info in the README at https://github.com/tmrowco/electricitymap-contrib/tree/master/mobileapp ?

I tried now but it seems I need a `google-services.json` file to be able to build it:

```bash
$ cordova prepare
...
Successfully set up for cordova
🎉  Successfully set up Sentry for your project 🎉
Source path does not exist: google-services.json

$ cordova build android
Source path does not exist: google-services.json
```

My Cordova version is `9.0.0 (cordova-lib@9.0.1)`.

corradio commented:
hmm - can you try having an empty file?

On Tue, Apr 14, 2020 at 7:32 PM Filip Barl <notifications@github.com> wrote:

> Have you tried building the cordova app by following the info in the
> README at
> https://github.com/tmrowco/electricitymap-contrib/tree/master/mobileapp ?
>
> I tried now but it seems I need a google-services.json file to be able to
> build it:
>
> $ cordova prepare
>
> ...
>
> Successfully set up for cordova
>
> 🎉  Successfully set up Sentry for your project 🎉
>
> Source path does not exist: google-services.json
>
>
>
> $ cordova build android
>
> Source path does not exist: google-services.json
>
> My Cordova version is 9.0.0 (cordova-lib@9.0.1).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tmrowco/electricitymap-contrib/pull/2317#issuecomment-613577721>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMUIKEVEDJNQIKG3EWNODDRMSM3JANCNFSM4LTS3OPA>
> .
>


-- 

Olivier Corradi

Founder, CEO

<https://www.linkedin.com/in/oliviercorradi> <https://twitter.com/corradio>

tmrow.com

fbarl commented:
> hmm - can you try having an empty file?

Indeed that worked :facepalm: 

I had to add empty files `google-services.json` and `GoogleService-Info.plist` to get past those errors but then I got:

```bash
$ cordova prepare
Cannot find module '../../src/plugman/platforms/ios'
Require stack:
- /home/fbarl/.nvm/versions/node/v12.9.1/lib/node_modules/cordova/node_modules/cordova-lib/src/hooks/Context.js
- /home/fbarl/.nvm/versions/node/v12.9.1/lib/node_modules/cordova/node_modules/cordova-lib/src/hooks/HooksRunner.js
- /home/fbarl/.nvm/versions/node/v12.9.1/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/install.js
- /home/fbarl/.nvm/versions/node/v12.9.1/lib/node_modules/cordova/node_modules/cordova-lib/src/plugman/plugman.js
- /home/fbarl/.nvm/versions/node/v12.9.1/lib/node_modules/cordova/node_modules/cordova-lib/cordova-lib.js
- /home/fbarl/.nvm/versions/node/v12.9.1/lib/node_modules/cordova/src/help.js
- /home/fbarl/.nvm/versions/node/v12.9.1/lib/node_modules/cordova/src/cli.js
- /home/fbarl/.nvm/versions/node/v12.9.1/lib/node_modules/cordova/bin/cordova
```

Searching around internet, I found https://github.com/nordnet/cordova-universal-links-plugin/issues/131, which hasn't been addressed as `cordova-universal-links` plugin that we're using seems to be deprecated.

A quick solution for me might be to downgrade Cordova though - @corradio which version are you using?

corradio commented:
I'm on cordova 8.0.0

fbarl commented:
Unfortunately I still couldn't get past `cordova build android` successfully :disappointed: These are the steps I went through (most of which seemed necessary):

* `npm install -g cordova@8.0.0` to make sure we're using the same version
* Add empty files `google-services.json` and `GoogleService-Info.plist` to the `/mobileapp` dir
* Apparently my Android SDK wasn't set up fully so I had to update the `ANDROID_HOME` env variable and [sign all the licences](https://stackoverflow.com/questions/39760172/you-have-not-accepted-the-license-agreements-of-the-following-sdk-components)
* After that I was hitting the `Cannot find module '../../src/plugman/platforms/ios'` error, which I got around by [patching this piece of code](https://github.com/nordnet/cordova-universal-links-plugin/issues/131#issuecomment-387761895) in the deprecated `cordova-universal-links-plugin`
* After that I got `Execution failed for task ':transformClassesWithDexBuilderForDebug'` which I got around by [fixing the JDK versions](https://github.com/flutter/flutter/issues/25271#issuecomment-449660007) to `1.8` in `bundle.gradle`

Then the next error I got was `Execution failed for task ':transformClassesWithDesugarForDebug'` and that's when I gave up as none of the solutions I found online worked for me.

It seemed that with each step above I'd get closer to making it buildable, but on the other hand the number of issues I encountered is just weird :thinking: Maybe it has to do with me doing it on Linux?

fbarl commented:
@corradio Ok, I think I found a solution in https://github.com/tmrowco/electricitymap-contrib/pull/2317/commits/9320cb115b7436ada5ca9ff21eaa0178e6ceb9be - I really hope Cordova builds fine after this change :crossed_fingers: 

The problem was that absolute resource routes were required to work in a default (`BrowserHistory`) setting of React Router but they don't play well with the `file://` protocol where `index.html` can be mounted at arbitrary locations.

What I did was to branch into 2 cases:

* For Cordova, I used `HashHistory` which works well with relative routes as the paths are stored in the URL hash, so everything technically happens at the root `/` - that would normally appear hacky in web browsers, but it doesn't matter much when the URLs are hidden as the behavior is basically the same
* For web, I kept using `BrowserHistory` with absolute routes so the behavior should be exactly the same

The branching of the history instance type was literally a one-line change, the main thing that increases complexity is the newly introduced `resolvePath` global helper which needs to wrap every image resource path now :/ (not including the paths from CSS files which I reverted to relative paths as this doesn't seem to interfere with React Router for some reason).

I tested it again in the web setting and also by opening the HTML pages generated by `./mobileapp/build.sh` and things look promising! :smiley:

corradio commented:
works!

fbarl commented:
Thanks for helping out with the PR @corradio!

I'll merge now as it seems to be working fine on all platforms, but just to be on the safe side I'd like to make sure I'm around when we test / bump the mobile versions to verify that nothing is broken in the release.
